README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ArchaeoPhases

<!-- badges: start -->

[![R-CMD-check](https://github.com/nfrerebeau/ArchaeoPhases/workflows/R-CMD-check/badge.svg)](https://github.com/nfrerebeau/ArchaeoPhases/actions)

[![CRAN
Version](http://www.r-pkg.org/badges/version/ArchaeoPhases)](https://cran.r-project.org/package=ArchaeoPhases)
[![CRAN
checks](https://cranchecks.info/badges/worst/ArchaeoPhases)](https://cran.r-project.org/web/checks/check_results_ArchaeoPhases.html)
[![CRAN
Downloads](http://cranlogs.r-pkg.org/badges/ArchaeoPhases)](https://cran.r-project.org/package=ArchaeoPhases)

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)

[![DOI
JSS](https://img.shields.io/badge/JSS-10.18637/jss.v093.c01-brightgreen)](https://doi.org/10.18637/jss.v093.c01)
<!-- badges: end -->

**This repository contains a heavily modified fork of the
*ArchaeoPhases* package. You might be interested in the [original
package](https://github.com/ArchaeoStat/ArchaeoPhases).**

## Overview

Statistical analysis of archaeological dates and groups of dates.
**ArchaeoPhases** allows to post-process Markov Chain Monte Carlo (MCMC)
simulations from [ChronoModel](https://chronomodel.com),
[Oxcal](https://c14.arch.ox.ac.uk/oxcal.html) or
[BCal](https://bcal.shef.ac.uk). This package provides functions for the
study of rhythms of the long term from the posterior distribution of a
series of dates (tempo and activity plot). It also allows the estimation
and visualization of time ranges from the posterior distribution of
groups of dates (e.g. duration, transition and hiatus between successive
phases).


    To cite ArchaeoPhases in publications use:

      Philippe, Anne & Vibet, Marie-Anne (2020). Analysis of Archaeological
      Phases Using the R Package ArchaeoPhases. Journal of Statistical
      Software, Code Snippets, 93(1), 1--25. DOI 10.18637/jss.v093.c01.

    Une entrée BibTeX pour les utilisateurs LaTeX est

      @Article{,
        title = {Analysis of Archaeological Phases Using the {R} Package {ArchaeoPhases}},
        author = {Anne Philippe and Marie-Anne Vibet},
        year = {2020},
        journal = {Journal of Statistical Software, Code Snippets},
        volume = {93},
        number = {1},
        page = {1--25},
        doi = {10.18637/jss.v093.c01},
      }

## Installation

You can install the released version of **ArchaeoPhases** from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("ArchaeoPhases")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("nfrerebeau/ArchaeoPhases")
```

## Usage

``` r
## Load packages
library(ArchaeoPhases)
```

Import a CSV file containing a sample from the posterior distribution:

``` r
## Read output from ChronoModel
path <- "chronomodel/output/"

## Events
path_events <- system.file(path, "Chain_all_Events.csv", package = "fasti")
(chrono_events <- read_chronomodel_events(path_events))
#> <EventsMCMC>
#> - Number of events: 16
#> - Calendar: CE

path_phases <- system.file(path, "Chain_all_Phases.csv", package = "fasti")
(chrono_phases <- read_chronomodel_phases(path_phases))
#> <PhasesMCMC>
#> - Modelled phases: EPI UP Ahmarian IUP
#> - Calendar: CE
```

**ArchaeoPhases** uses
[**ggplot2**](https://github.com/tidyverse/ggplot2) for plotting
information. This makes it easy to customize diagrams (e.g. using themes
and scales).

### Analysis of a series of dates

``` r
## Plot the first event
plot(chrono_events, select = 1, interval = "hpdi")

## Plot all events
plot(chrono_events)
#> Picking joint bandwidth of 49.2
```

<img src="man/figures/README-events-plot-1.png" width="50%" /><img src="man/figures/README-events-plot-2.png" width="50%" />

``` r
## Tempo plot
tp <- tempo(chrono_events, level = 0.95)
plot(tp)

## Activity plot
ac <- activity(chrono_events)
plot(ac)
```

<img src="man/figures/README-tempo-plot-1.png" width="50%" /><img src="man/figures/README-tempo-plot-2.png" width="50%" />

### Analysis of a group of dates (phase)

``` r
boundaries(chrono_phases, level = 0.95)
#>           start    end
#> EPI      -28979 -26970
#> UP       -38570 -29369
#> Ahmarian -42168 -37433
#> IUP      -43240 -41161
```

``` r
plot(chrono_phases)
```

<img src="man/figures/README-phases-plot-1.png" style="display: block; margin: auto;" />

``` r
## Set chronological order
## (from the oldest to the youngest phase)
set_order(chrono_phases) <- c("IUP", "Ahmarian", "UP", "EPI")
get_order(chrono_phases)
#> [1] "IUP"      "Ahmarian" "UP"       "EPI"
```

``` r
plot(chrono_phases, level = 0.95)
```

<img src="man/figures/README-succession-plot-1.png" style="display: block; margin: auto;" />


NEWS
NEWS
==============
ArchaeoPhases 1.6 (Release date: )
==============

* New functions for Allen's interval algebra: allen_analyze(), allen_joint_concurrency(), allen_observe_frequency(), allen_illustrate(), allen_observe().

* Fixed a bug that failed to export the reproduce() function.

* Added a vignette on object reproducibility.

==============
ArchaeoPhases 1.5 (Release date: )
==============

* Fixed a bug in MultiHPD() that ignored the roundingOfValue parameter.

* Fixed a bug in MarginalStatistics() that triggered an error if the function was passed a constant MCMC chain.

* New read functions for MCMC data: read_bcal(), read_oxcal(), read_chronomodel().
  * The new read functions return S3 objects that can identify the file that produced them.
  * The new read functions are built on read_csv(), which can read data from a file, connection, or the clipboard.

* New plot functions: multi_dates_plot(), tempo_activity_plot(), tempo_plot(), marginal_plot(), multi_marginal_plot(), and occurrence_plot().
  * The new plot functions are functional replacements for the originals with camelCase names, e.g., TempoPlot() -> tempo_plot().
  * The new plot functions return S3 objects with plot() and reproduce() methods.
  * The S3 objects inherit from data frame and can be passed to appropriate statistical functions to summarize the data in the plot.

* New statistical functions: credible_interval(), multi_credible_interval(), multi_hpd(), dates_hiatus(), phases_gap(), marginal_statistics(), and phase_statistics().
  * The new statistical functions are functional replacements for the originals with camelCase names, e.g., CredibleInterval() -> credible_interval().
  * The new statistical functions return lists in canonical R fashion.
  * The phase_statistics() function is augmented with a round_to parameter.

* New statistical function multi_marginal_statistics().

* New statistical function estimate_ranges() that can be used to estimate the sensitivity of calibration results to different model parameters.

==============
ArchaeoPhases 1.4 (Release date: 2017-12-12)
==============

* Includes an update of the TempoPlot -> Optimisation of the credible intervals as already done in OccurrencePlot

* Includes a minor update of the vignettes

* Includes an update of the function MarginalPlot and adds a new function : MultiMarginalPlot()

* Includes an update of the function MarginalStatistics and adds a new function : MultiMarginalMarginalStatistics()

* Includes an update of the shiny web application (called by function app_ArchaeoPhases()) that did not work in the previous version



==============
ArchaeoPhases 1.3 (Release date: 2017-09-05)
==============

* Includes an update of the function ImportCSV and a new function for BCal users called ImportCSV.BCal

* Includes an update of the MultiDatesPlot. The graphic is now done with ggplot2

* Includes an update of the TempoPlot and TempoActivityPlot functions

* Includes a new function : OccurrencePlot

* Includes an update of the vignette that describes the main functions of the package

* Includes an update of the shiny web application (called by function app_ArchaeoPhases())


==============
ArchaeoPhases 1.2 (Release date: 2017-06-12)
==============

* Includes an update of the vignette that describes the main functions of the package

* Includes an update of the shiny web application (called by function app_ArchaeoPhases())


==============
ArchaeoPhases 1.1 (Release date: 2017-03-28)
==============

* Includes a vignette that describes the main functions of the package

* Includes an update of the ImportCSV function in order to import the raw MCMC generated by 'BCal' and to convert the MCMC samples from the date format cal BP (in years before 1950) to the date format BC/AD

* Includes a new dataset of MCMC samples generated by BCal : Fishpond.RData

* Includes an update of the coda.mcmc function

* Includes an update of the TempoPlot function using the package 'ggplot2'

* Includes an update of all graphic functions : it is now possible to export all graphics and to choose the colors associated to the characteristics of groups of dates.

* Includes an update of the shiny web application (called by function app_ArchaeoPhases())

==============
ArchaeoPhases 1.0 (Release date: 2017-02-06)
==============

* Initial release

* Includes all functions of RChronoModel and their updates

* Includes a function coda.mcm() that creates a MCMC_list in order to use the package 'coda'

* Includes a shiny application and the function app_ArchaeoPhases() to call it from R

* RChronoModel is now obsolete and replaced by ArchaeoPhases


R/plot.R
# PLOT
#' @include AllClasses.R AllGenerics.R
NULL

# MCMC =========================================================================
#' @export
#' @method autoplot MCMC
autoplot.MCMC <- function(object, ..., select = NULL, groups = NULL,
                          density = TRUE, interval = NULL, level = 0.95,
                          decreasing = TRUE) {
  ## Calendar scale
  gg_x_scale <- scale_calendar(get_calendar(object))

  ## Select data
  if (!is.null(select)) {
    object <- object[, select, drop = FALSE]
  }

  ## Group data
  grp <- data.frame(id = character(0), Group = character(0))
  aes_grp <- NULL
  if (!is.null(groups)) {
    if (length(groups) != ncol(object)) {
      msg <- "%s must be of length %d, not %d."
      stop(sprintf(msg, sQuote("groups"), ncol(object), length(groups)),
           call. = FALSE)
    }
    grp <- data.frame(id = names(object), Group = as.character(groups))
    aes_grp <- ggplot2::aes(color = .data$Group, fill = .data$Group)
  }

  ## Reorder data
  decreasing <- ifelse(is_BP(object), !decreasing, decreasing)
  data <- reorder(object, decreasing = decreasing)

  ## Compute interval
  gg_inter <- NULL
  if (!is.null(interval)) {
    interval <- match.arg(interval, choices = c("credible", "hpdi"))
    fun <- switch (
      interval,
      credible = credible,
      hpdi = hpdi,
      stop(sprintf("There is no such interval: %s.", interval), call. = FALSE)
    )
    inter <- fun(data, level = level)
    inter <- bind_intervals(inter)
    inter$id <- factor(inter$id, levels = unique(inter$id))

    ## Add group (if any)
    inter <- merge(inter, grp, by = "id", all.x = TRUE, all.y = FALSE,
                   sort = FALSE)

    aes_inter <- ggplot2::aes(x = .data$lower, y = .data$id,
                              xend = .data$upper, yend = .data$id)
    gg_inter <- ggplot2::geom_segment(mapping = aes_inter, data = inter,
                                      size = 2)
  }

  ## Density
  gg_dens <- NULL
  if (density) {
    if (ncol(data) > 1) {
      ## Build long table for ggplot2
      dens <- data.frame(
        row = as.vector(row(data, as.factor = FALSE)),
        column = as.vector(col(data, as.factor = T)),
        value = as.vector(data),
        stringsAsFactors = FALSE
      )
      ## Keep original ordering
      dens$column <- as.character(dens$column)
      dens$id <- factor(dens$column, levels = unique(dens$column))

      aes_dens <- ggplot2::aes(
        x = .data$value,
        y = .data$id,
        group = .data$id
      )
      gg_dens <- ggridges::geom_density_ridges(
        mapping = aes_dens,
        data = dens,
        panel_scaling = TRUE,
        rel_min_height = 0.01,
        scale = 1.75,
        alpha = 0.5
      )
    } else {
      data <- as.vector(data)
      ## Compute density
      d <- stats::density(data, n = getOption("ArchaeoPhases.grid"))
      dens <- data.frame(
        x = d$x,
        y = d$y
      )

      aes_dens <- ggplot2::aes(
        x = .data$x,
        y = .data$y
      )
      gg_dens <- ggplot2::geom_path(
        mapping = aes_dens,
        data = dens
      )

      if (!is.null(interval)) {
        ## Is credible?
        cred <- is_credible(d$x, inter[, c(1, 2)])
        int <- dens[which(cred), , drop = FALSE]
        int$Interval <- paste0(round(level * 100), "%")

        aes_inter <- ggplot2::aes(
          x = .data$x,
          y = .data$y,
          fill = .data$Interval
        )
        gg_inter <- ggplot2::geom_area(
          mapping = aes_inter,
          data = int
        )
      }
    }
  }

  ## ggplot2
  ggplot2::ggplot() +
    aes_grp +
    gg_dens +
    gg_inter +
    gg_x_scale
}

#' @export
#' @rdname plot
#' @aliases autoplot,MCMC-method
setMethod("autoplot", "MCMC", autoplot.MCMC)

#' @export
#' @method plot MCMC
plot.MCMC <- function(x, select = NULL, groups = NULL, density = TRUE,
                      interval = NULL, level = 0.95, decreasing = TRUE, ...) {
  gg <- autoplot(object = x, ..., select = select, groups = groups,
                 density = density, interval = interval, level = level,
                 decreasing = decreasing) +
    ggplot2::guides(fill = ggplot2::guide_legend(ncol = 2)) +
    ggplot2::theme_bw() +
    ggplot2::theme(axis.title.y = ggplot2::element_blank())
  print(gg)
  invisible(x)
}

#' @export
#' @describeIn plot Plots of credible intervals or HPD regions of a series of
#' events.
#' @aliases plot,MCMC,missing-method
setMethod("plot", c(x = "MCMC", y = "missing"), plot.MCMC)

# PhasesMCMC ===================================================================
#' @export
#' @method autoplot PhasesMCMC
autoplot.PhasesMCMC <- function(object, ..., level = 0.95, decreasing = TRUE,
                                succession = is_ordered(object), facet = TRUE) {
  ## Calendar scale
  gg_x_scale <- scale_calendar(get_calendar(object))

  if (succession) {
    gg_phases <- plot_succession(object, level = level,
                                 decreasing = decreasing)
  } else {
    gg_phases <- plot_density(object, level = level, decreasing = decreasing,
                              ..., facet = facet)
  }

  ## ggplot2
  ggplot2::ggplot() +
    gg_phases +
    gg_x_scale
}

#' @export
#' @rdname plot
#' @aliases autoplot,PhasesMCMC-method
setMethod("autoplot", "PhasesMCMC", autoplot.PhasesMCMC)

#' @export
#' @method plot PhasesMCMC
plot.PhasesMCMC <- function(x, level = 0.95, decreasing = TRUE,
                            succession = is_ordered(x), facet = TRUE, ...) {
  gg <- autoplot(object = x, ..., level = level, decreasing = decreasing,
                 succession = succession, facet = facet) +
    ggplot2::theme_bw() +
    ggplot2::theme(axis.title.y = ggplot2::element_blank())
  print(gg)
  invisible(x)
}

#' @export
#' @describeIn plot Plots the characteristics of a group of events (phase).
#' @aliases plot,PhasesMCMC,missing-method
setMethod("plot", c(x = "PhasesMCMC", y = "missing"), plot.PhasesMCMC)

# Helpers ======================================================================
#' @param x A [`PhasesMCMC`] object.
#' @return A \pkg{ggplot2} layer.
#' @noRd
plot_succession <- function(x, level = 0.95, decreasing = TRUE,
                            size = 2, fill = "grey50", alpha = 0.5) {
  ## Time range
  duree <- boundaries(x, level = level)
  ord <- rank(duree$start)
  duree$rank <- if (decreasing) -ord else ord
  pha <- get_order(x)
  duree$Phase <- factor(pha, levels = pha, ordered = TRUE)

  gg_trans <- gg_hiatus <- NULL
  if (is_ordered(x)) {
    ## Transition
    trans <- transition(x, level = level)
    trans <- data.frame(
      xmin = trans$start,
      xmax = trans$end,
      ymin = min(duree$rank) - 0.5,
      ymax = max(duree$rank) + 0.5,
      labels = rownames(trans)
    )

    ## Hiatus
    hia <- hiatus(x, level = level)
    hia <- data.frame(
      xmin = hia$start,
      xmax = hia$end,
      ymin = min(duree$rank) - 0.5,
      ymax = max(duree$rank) + 0.5,
      labels = rownames(hia)
    )
    hia <- stats::na.omit(hia)

    ## Layers
    aes_rect <- ggplot2::aes(
      xmin = .data$xmin,
      xmax = .data$xmax,
      ymin = .data$ymin,
      ymax = .data$ymax
    )
    if (nrow(trans) > 0) {
      gg_trans <- ggplot2::geom_rect(
        mapping = aes_rect,
        data = trans,
        fill = fill,
        alpha = alpha
      )
    }
    if (nrow(hia) > 0) {
      gg_hiatus <- ggplot2::geom_rect(
        mapping = aes_rect,
        data = hia,
        fill = fill,
        alpha = alpha + 0.25
      )
    }
  }

  ## Layer
  aes_range <- ggplot2::aes(
    x = .data$start,
    y = .data$rank,
    xend = .data$end,
    yend = .data$rank,
    color = .data$Phase
  )
  gg_range <- ggplot2::geom_segment(
    mapping = aes_range,
    data = duree,
    size = size
  )
  gg_y_scale <- ggplot2::scale_y_continuous(
    name = "Phases",
    breaks = duree$rank,
    labels = duree$Phase
  )
  list(gg_trans, gg_hiatus, gg_range, gg_y_scale)
}

#' @param x A [`PhasesMCMC`] object.
#' @return A \pkg{ggplot2} layer.
#' @noRd
plot_density <- function(x, level = 0.95, decreasing = TRUE, ..., facet = TRUE,
                         color = "black", size = 2, alpha = 0.5) {
  ## Get phases
  pha <- as.list(x)

  ## Density
  n <- getOption("ArchaeoPhases.grid")
  dens <- lapply(
    X = pha,
    FUN = function(x, n, ...) {
      a <- stats::density(x[, , 1, drop = TRUE], n = n, ...)
      b <- stats::density(x[, , 2, drop = TRUE], n = n, ...)
      data.frame(
        x = c(a$x, b$x),
        y = c(a$y, b$y),
        z = rep(c("Begin", "End"), each = n)
      )
    },
    n = n, ...
  )

  ## Time range
  duree <- boundaries(x, level = level)
  ord <- rank(duree$start)
  duree$rank <- if (decreasing) -ord else ord
  duree_phase <- rownames(duree)
  duree$Phase <- factor(duree_phase, levels = duree_phase,
                        ordered = is_ordered(x))
  # duree$Range <- paste0(round(level * 100, digits = 0), "%")

  ## Adjust y position
  y_max <- vapply(X = dens, FUN = function(x) max(x$y), FUN.VALUE = numeric(1))
  if (facet) {
    duree$y <- y_max * 1.1
  } else {
    y_duree <- max(y_max) * (1 + duree$rank / 10)
    duree$y <- if (decreasing) y_duree + diff(range(y_max)) else y_duree
  }

  ## Bind densities
  dens <- do.call(rbind, dens)
  dens_phase <- names(pha)
  dens$Phase <- factor(rep(dens_phase, each = 2 * n), levels = dens_phase,
                       ordered = is_ordered(x))
  dens$Boundary <- factor(dens$z, levels = c("Begin", "End"), ordered = TRUE)

  ## Layer
  gg_facet <- NULL
  if (facet) {
    gg_facet <- ggplot2::facet_grid(
      rows = ggplot2::vars(.data$Phase),
      scales = "free_y"
    )
  }
  aes_dens <- ggplot2::aes(
    x = .data$x,
    ymin = 0,
    ymax = .data$y,
    fill = .data$Phase,
    linetype = .data$Boundary
  )
  gg_dens <- ggplot2::geom_ribbon(
    mapping = aes_dens,
    data = dens,
    color = color,
    alpha = alpha
  )
  aes_range <- ggplot2::aes(
    x = .data$start,
    y = .data$y,
    xend = .data$end,
    yend = .data$y,
    color = .data$Phase
  )
  gg_range <- ggplot2::geom_segment(
    mapping = aes_range,
    data = duree,
    size = size
  )
  gg_y_scale <- ggplot2::scale_y_continuous(name = "Density")
  list(gg_dens, gg_range, gg_facet, gg_y_scale)
}
